### ISSUE TITLE
Crash with custom-url-scheme on android

### ISSUE THREAD

**Issue by jlchereau** [EddyVerbruggen/cordova-plugin-safariviewcontroller#51]
I am trying to replace InAppBrowser with SafariViewController in oAuth flow involving a web server. See #34.

The redirect_uri on the web server (which is whitelisted with the oAuth provider) returns:

```
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <meta http-equiv="Content-Security-Policy" content="default-src 'none';script-src 'unsafe-inline';">
</head>
<body>
    <script>window.location.assign('com.example.app://token' + window.location.hash);</script>
</body>
</html>
```

then the handleOpenURL event handler parses the oAuth token in the hash.

This works perfectly in iOS but in android SafariViewController crashes on the above page before hitting handleOpenURL.

I know the custom url scheme is properly configured because I can hit handleOpenUrl from a link in a page opened in Chrome.

if I replace:

```
<script>window.location.assign('com.example.app://token' + window.location.hash);</script>
```

with

```
<script>window.location.assign('http://www.google.com');</script>
```

the redirection works just fine so SafariViewController simply does not like 'com.example.app://' on Android only, not iOS.

Any suggestions to have SafariViewController (Chrome Tabs) accept a custom scheme made with https://github.com/EddyVerbruggen/Custom-URL-scheme on Android ?

Project can be found at https://github.com/kidoju/Kidoju-Mobile.

I am using:
- Phonegap 6.3.0
- Android Platform 5.2.1
- Latest versions of all plugins
(latest versions on Phonegap build as of today)

I am testing on a Nexus 7 with Android 5.1.1 and Chrome 54.0.2840.85.

**Comment by yakirn** [comment]
I'm not sure... Could it be that since your activity (com.example.app)
launches the chrome custom tabs activity, it's not possible for the
"launchee" to launch the launcher again before it is terminated?
Try to see if there's a way to close the custom tabs activity and then
launch yours again.

On Sat, 3 Dec 2016 at 17:29 Jacques L. Chereau <notifications@github.com>
wrote:

I am trying to replace InAppBrowser with SafariViewController in oAuth flow
involving a web server.

The redirect_uri returns:

<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <meta http-equiv="Content-Security-Policy" content="default-src
'none';script-src 'unsafe-inline';">
</head>
<body>
    <script>window.location.assign('com.example.app://token' +
window.location.hash);</script>
</body>
</html>

then the handleOpenURL event handler parses the oAuth token in the hash.

This works perfectly in iOS but in android SafariViewController crashes on
the above page before hitting handleOpenURL.

if I replace:

<script>window.location.assign('com.example.app://token' +
window.location.hash);</script>

with

<script>window.location.assign('http://www.google.com');</script>

the redirection works just fine so SafariViewController (android only, not
iOS) simply does not like 'com.example.app://'.

Any suggestions?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/EddyVerbruggen/cordova-plugin-safariviewcontroller/issues/51>,
or mute the thread

**Comment by jlchereau** [comment]
First this is an Android (Chrome Tabs) only issue. It does not affect iOS (SFSafariViewController) which work fine.

After further investigations, handleOpenURL is hit:

```
window.handleOpenURL = function (url) {
    setTimeout(function() {
        alert(url);
        // doSomethingWith(url);
    }, 0);
});
```

In this case:
1) SafariViewController (Chrome Tabs) closes without the hide method being called (crash?);
2) The phonegap SplashScreen is displayed instead of getting back to the web view;
3) After about 10sec the alert shows the url and the WebView is displayed again, as if the webview had crashed and reloaded.

**Comment by yakirn** [comment]
That's because in iOS the SafariViewController run as part of the application. While in android it is a different Activity (like a process) that is activated by your Cordova activity. What I'm saying is that it's completely different code that runs on a completely different framework, so what is and isn't working on iOS is not an indication to anything IMO.
I hope I'll have time to take a better look over the weekend. If you can find a working example of authentication using Chrome custom tabs, it might speed things up.
Sorry I can't help you more, and good luck!

**Comment by jlchereau** [comment]
Thank you. You can easily reproduce this by:

1. Configuring a phonegap application with:
    - https://github.com/EddyVerbruggen/Custom-URL-scheme
    - https://github.com/EddyVerbruggen/cordova-plugin-safariviewcontroller
2: Binding the click event of a button to

```
window.SafariViewController.show({  url: 'http://10.0.0.105:8080/temp/scheme.html' });
```

where scheme.html is:

```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
</head>
<body>
    <h1><a href="com.example.app://token/#auth_token=ttttttttttt&state=ssssssss">Authenticate</a></h1>
</body>
</html>
```

**Comment by jlchereau** [comment]
This might help:
- http://stackoverflow.com/questions/36084681/chrome-custom-tabs-redirect-to-android-app-will-close-the-app

**Comment by samuelgodbout** [comment]
I am having the exact same issue as @jlchereau. The handleOpenUrl does not get hit while calling the custom url scheme in the chrome custom tab

**Comment by jlchereau** [comment]
I have considered the fact that my app and the Chrome Custom Tabs are different activities with a caller and a callee and I have therefore tried to modify the android launchMode of my application (standard, singleInstance, singleTask, singleTop) but this makes no difference.

**Comment by EtienneLem** [comment]
@jlchereau Have you found a workaround?

The only way I managed to make Google OAuth work is by using a good old `window.open` **without** `inappbrowser`, so that it opens in Chrome browser and then send back data with the custom URL scheme.

**Comment by simonwatt** [comment]
I'm using PhoneGap Build and had a similar sounding issue that I resolved by adding the following to my config.xml:

Initially I had:
    <preference name="AndroidLaunchMode" value="singleTask" />

But issue didn't go away until I added:

```
 <gap:config-file platform="android" parent="/manifest/application">
        <activity android:launchMode="singleTask" />
   </gap:config-file>
```

And also changing widget element to look like:
```
<widget xmlns     = "http://www.w3.org/ns/widgets"
        xmlns:gap = "http://phonegap.com/ns/1.0"
        xmlns:android = "http://schemas.android.com/apk/res/android"
        id        = "[app id here]"
        version   = "[version here]">
```

**Comment by jlchereau** [comment]
Are you sure? I have tried all launchMode values (see my comment of 20 dec) on Android 5 and it did not work.

**Comment by simonwatt** [comment]
It's what made the difference for me anyway. I was following documentation which only had <preference name="AndroidLaunchMode" value="singleTask" /> in the config, and it was getting ignored. When I configured as per above (which I found via a stackoverflow comment) the issue went away.

This was on Android 4.4, PhoneGap version "cli-6.1.0".  I'll try it on Android 5 and latest version of PhoneGap supported by Build within the next few days though so will be curious if the issue is still there.
